cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/ProvisioningService.java

```diff
@@ -291,6 +351,79 @@ public void validateProperties(String provisionerName, Map<String, String> prope
+    ProvisionerContext context =
+      new DefaultProvisionerContext(programRunId, taskInfo.getProvisionerProperties(),
+                                    sparkCompat, createSSHContext(taskInfo.getSecureKeyInfo()));
+
+    // TODO: (CDAP-13246) pick up timeout from profile instead of hardcoding
+    ProvisioningTask task = new ProvisionTask(taskInfo, transactional, datasetFramework, provisioner, context,
+                                              provisionerNotifier, 300);
+    return () -> {
+      try {
+        task.execute();
+      } catch (InterruptedException e) {
+        LOG.debug("Provision task for program run {} interrupted.", taskInfo.getProgramRunId());
+      } catch (Exception e) {
+        LOG.info("Provision task for program run {} failed.", taskInfo.getProgramRunId(), e);
+      }
+    };
+  }
+
+  private Runnable createDeprovisionTask(ProvisioningTaskInfo taskInfo, Provisioner provisioner,
+                                         Consumer<ProgramRunId> taskCleanup) {
+    Map<String, String> properties = taskInfo.getProvisionerProperties();
+    ProvisionerContext context =
+      new DefaultProvisionerContext(taskInfo.getProgramRunId(), properties,
+                                    sparkCompat, createSSHContext(taskInfo.getSecureKeyInfo()));
+    DeprovisionTask task = new DeprovisionTask(taskInfo, transactional, datasetFramework, 300,
+                                               provisioner, context, provisionerNotifier, locationFactory);
+    return () -> {
+      try {
+        task.execute();
+        taskCleanup.accept(taskInfo.getProgramRunId());
+      } catch (InterruptedException e) {
+        // We can get interrupted if the task is cancelled or CDAP is stopped. In either case, just return.
+        // If it was cancelled, state cleanup is left to the caller. If it was CDAP master stopping, the task
+        // will be resumed on master startup
+        LOG.debug("Deprovision task for program run {} interrupted.", taskInfo.getProgramRunId());
+      } catch (Exception e) {
+        // Otherwise, if there was an error deprovisioning, run the cleanup
+        LOG.info("Deprovision task for program run {} failed.", taskInfo.getProgramRunId(), e);
+        taskCleanup.accept(taskInfo.getProgramRunId());
+      }
+    };
+  }
+
+  private List<ProvisioningTaskInfo> getInProgressTasks() {
+    return Retries.callWithRetries(
+      () -> Transactionals.execute(transactional, dsContext -> {
+        ProvisionerDataset provisionerDataset = ProvisionerDataset.get(dsContext, datasetFramework);
+        return provisionerDataset.listTaskInfo();
+      }),
+      co.cask.cdap.common.service.RetryStrategies.exponentialDelay(10, 120, TimeUnit.SECONDS),
+      t -> {
+        // don't retry if we were interrupted, or if the service is not running
+        State serviceState = state();
+        if (serviceState != State.STARTING && serviceState != State.RUNNING) {
```

Comment:
Will the state be `RUNNING`? this method only gets call from the `startUp` method and block until finished, right?